src/Domain/Encryption/Crypter.ts

```diff
@@ -0,0 +1,59 @@
+import { inject, injectable } from 'inversify'
+import { randomBytes, pbkdf2Sync, createCipheriv, createDecipheriv } from 'crypto'
+
+import TYPES from '../../Bootstrap/Types'
+import { CrypterInterface } from './CrypterInterface'
+
+@injectable()
+export class Crypter implements CrypterInterface {
+  static readonly ENCRYPTION_VERSION_1 = 1
+
+  private readonly ENCRYPTION_KEY_LENGTH = 32
+
+  constructor(
+    @inject(TYPES.ENCRYPTION_ITERATIONS) private encryptionIterations: number
+  ) {
+  }
+
+  async encrypt(encryptionVersion: number, data: string, masterKey: string): Promise<string> {
+    if (encryptionVersion !== Crypter.ENCRYPTION_VERSION_1) {
+      throw new Error(`Supported versions of encryption: ${Crypter.ENCRYPTION_VERSION_1}`)
+    }
+
+    const iv = randomBytes(16)
+
+    const salt = randomBytes(64)
+
+    const key = pbkdf2Sync(masterKey, salt, this.encryptionIterations, this.ENCRYPTION_KEY_LENGTH, 'sha512')
+
+    const cipher = createCipheriv('aes-256-gcm', key, iv)
+
+    const encrypted = Buffer.concat([cipher.update(data, 'utf8'), cipher.final()])
+
+    const tag = cipher.getAuthTag()
+
+    return Buffer.concat([salt, iv, tag, encrypted]).toString('base64')
+  }
+
+  async decrypt(encryptionVersion: number, data: string, masterKey: string): Promise<string> {
+    if (encryptionVersion !== Crypter.ENCRYPTION_VERSION_1) {
+      throw new Error(`Supported versions of encryption: ${Crypter.ENCRYPTION_VERSION_1}`)
+    }
+
+    const bData = Buffer.from(data, 'base64')
+
+    const salt = bData.slice(0, 64)
```

Comment:
Why not use a character based separator instead? These numbers are so..arbitrary. If you instead separated components by a `:` character, you could just do `bData.split(':')` and then read components like const `salt = components[0[, iv = components[1]`, etc. Seems much cleaner.

Reply:
I'll change this, but out of curiosity isn't that join with ':' a vulnerability? That way if the key get's into wrong hands a perpetrator can clearly see where all of the key parts start and end in particular the salt ... I think the purpose of the salt is to obfuscate (with it's unknown length) the start of the encrypted data. This defeats the purpose of having a salt alltogether actually.

Reply:
No, this isn't the case at all. A salt is simply for making it so that encrypting the same data twice yields different results. Unlike a hash function no two encrypted payloads should ever be alike, otherwise that's a huge leak of information.

> To make it hard to guess where the salt ends and encrypted data starts

Our code is open source. You're not going to fool anyone 😉 So it's better to optimize for clarity than believing you're going to trick a sophisticated hacker about the length of a salt. Only person you'll be confusing is likely the next developer working on this project.

Reply:
> Our code is open source. You're not going to fool anyone 😉 

it's not my intention to fool anyone. If the length is an env var value then by any means is this open to public and adds additional layer of security.

> Only person you'll be confusing is likely the next developer working on this project.

What is confusing about the `encryptionSaltLength` variable that comes from env vars ?

Reply:
> adds additional layer of security

It doesn't. Salts are meant to be treated as public information. In some algorithms salt lengths are actually required to be fixed.

Reply:
with `AES-256-GCM` salt is not fixed thus this can add another layer of obscurity to decoding the encrypted data.

Reply:
neither is the IV length fixed with this

Reply:
There's no need to "layer" encryption. 1 layer is enough. That's the encryption part. Salts/ivs are public info. Good salts should be hard to predict (but can nonetheless still be the user's email address), and IVs should be cryptographically random, but are still public information.